File: projects/04/java/factorial.java
/* factorial.java is a driver for function factorial().
 * Author: Sarah Whitten
 * Date: February 25, 2019
 ****************************************************************/

import java.util.*;

public class factorial
{
    public static void main(String[] args)
    {
    	Scanner s = new Scanner(System.in);
    	System.out.print("To compute n!, enter n: ");

    	int base = s.nextInt();
    	int result = factorial(base);

    	System.out.println(base + "! = " + result);
    }

    private static int factorial(int base)
    {
    	int counter = 1;
    	for (int i = 2; i <= base; i++)
    	{
    		counter *= i;
    	}
    	return counter;
    }
}

File: projects/04/clojure/src/factorial.clj
;;;; factorial.clj is a driver for function factorialOf.
;;;;
;;;; Input: Integer
;;;; Output: Factorial of given integer
;;;;
;;;; Usage: clojure -m factorial
;;;;
;;;; Auhtor: Sarah Whitten
;;;; Date: February 25, 2019
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns factorial)                    ; name the program

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; displayLogTable() recursively generates the log-table.
;;; Receive: start the first value for the table;
;;;          stop, the last  value for the table;
;;;           step, the step value for the table;
;;; Precondition: start < stop && increment > 0.
;;; Output: The table of logs from start to stop, by step.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; direct-recursive version
(defn factorialOf [base limit i]
    (if (<= i limit)
    	(factorialOf (* base i) limit (+ i 1))
    	base)
)

(defn -main []
	(println "To compute n!, enter n: ")
	(let 
		[limit (double (read))]
	(print limit) (print "! = ")
	(println (factorialOf 1 limit 1))
	)
)
